### DISCUSSION TITLE
How to grep for a string only within a sub-folder?

### DISCUSSION THREAD

**Discussion by benfrain** [ibhagwan/fzf-lua#1501] [Q&A] [answered]
Suppose I want to search for the string "Hidden" but only within a sub-folder called 'MySubFolder', what syntax can I use to achieve that in the input box?

**Comment by ibhagwan** [comment]
```vim
:FzfLua grep search=hidden cwd=MySubFolder
```

**Comment by ibhagwan** [comment]
Unless you mean live_grep_glob, you’d need to look at the glob spec of ripgrep, should be something like `hidden -- mysubfolder/**`

**Reply by benfrain** [reply]
Yes, it's live grep I'm using and that doesn't seem to work, will take a look in ripgrep docs

**Reply by benfrain** [reply]
From the CLI I can do `rg "hidden" MySubFolder/` and it works but that does not work in the live grep of FZF 🤷‍♂️

**Reply by ibhagwan** [reply]
> Yes, it's live grep I'm using and that doesn't seem to work, will take a look in ripgrep docs

fzf-lua displays what the rg command outputs, use `:FzfLua live_grep debug=true` and look at the generated command, if the same command works in the CLI it will work in fzf-lua, also note that live_grep generates a glob after the separator `--` but once you switched to fuzzy matching the rg command remains static and you’re now doing fuzzy matching controlled by fzf:
![IMG_2604](https://github.com/user-attachments/assets/41bb1671-6f18-4e0d-8bb7-2100097cfc6b)

**Reply by benfrain** [reply]
Yes, my [debug] seems to not have the --iglob bit... could I be missing something in my config?

<img width="820" alt="image" src="https://github.com/user-attachments/assets/976fb0c7-6c32-4022-afb1-9121e69edbab">

**Reply by benfrain** [reply]
And this is the format I would write on the CLI which is also not producing anything in FZF but I think I'm probably misunderstanding (sorry!)

<img width="822" alt="image" src="https://github.com/user-attachments/assets/458d0ed8-d513-459a-bceb-f450fcd4b520">

**Reply by ibhagwan** [reply, marked as answer]
> Yes, my [debug] seems to not have the --iglob bit... could I be missing something in my config?

You either have to use `live_grep_glob` or set `grep = { rg_glob = true }` in your setup call to enable globs for all grep/live_grep pickers.

> And this is the format I would write on the CLI which is also not producing anything in FZF but I think I'm probably misunderstanding (sorry!)

As you can see from the generated command everything you type is searched for as a regex, this obviously leads to no results, what you’re writing in the shell is telling rg to search only the sub folder which is equivalent to:
```lua
:FzfLua grep search=leader cwd=setup
-- or live grep and input “leader”
:FzfLua live_grep cwd=setup
```
The latter is obviously different than using globs to limit the search to the folder, none is better than the other, it’s just two different tools at your disposal, if you know in advance which folder you wish to search use `cwd`, if you only find out after opening search which folder you want to focus, use globs.

**Reply by ibhagwan** [reply]
Just tested, once you enable `rg_glob` the `subfolder/**` syntax should work with `rg`, see the generated commands and the corresponding output:
> Search in tests subfolder excluding spec files
![image](https://github.com/user-attachments/assets/c989b681-3bb8-417c-b827-d1487dce75dd)

> Search anywhere but the tests folder and also exclude markdown files
![image](https://github.com/user-attachments/assets/55e70690-0835-45f3-a466-7fe7f6486302)

**Reply by benfrain** [reply]
> You either have to use live_grep_glob or set grep = { rg_glob = true } in your setup call to enable globs for all grep/live_grep pickers.

That was the bit I was missing, thanks! I can now do `"Leader" -- lua/setup/**`  👍

Is there any downside to having that in my default setup call (perf?)

**Reply by ibhagwan** [reply]
> Is there any downside to having that in my default setup call (perf?)

Not IMO, I have it enabled in my setup.

**Reply by benfrain** [reply]
Lovely, thanks for all your help and this incredible plugin 🙏

Do you have Kofi or paypal so I can buy you a virtual beer/beverage?

**Reply by ibhagwan** [reply]
> Lovely, thanks for all your help and this incredible plugin 🙏
> 
> Do you have Kofi or paypal so I can buy you a virtual beer/beverage?

Ty @benfrain for the kind words, appreciate the offer but that won’t be necessary 🙏
